DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/stp/test_stp_convergence.py

```diff
@@ -0,0 +1,185 @@
+        'type': 'bridge',
+        'vlan_filstering': 0,
+        'stp_state': 1} for bridge in bridges]}])
+    assert out[0][dent]['rc'] == 0, 'Failed to add bridge'
+
+    out = await Mstpctl.add(input_data=[{dent: [{'bridge': bridge} for bridge in bridges]}])
+    assert out[0][dent]['rc'] == 0, 'Failed to add bridge with mstpd'
+
+    out = await Mstpctl.set(input_data=[{dent: [{
+        'bridge': bridge,
+        'parameter': 'forcevers',
+        'version': version} for bridge in bridges]}])
+    assert out[0][dent]['rc'] == 0, 'Failed to set stp/rstp version'
+
+    # 2. Enslave ports to bridges
+    out = await IpLink.set(input_data=[{dent: [{
+        'device': port,
+        'operstate': 'down'} for port in loopback_ports.values()]}])
+    assert out[0][dent]['rc'] == 0, 'Failed setting links to state down'
+
+    for bridge, ports in bridges.items():
+        out = await IpLink.set(input_data=[{dent: [{'device': port, 'master': bridge} for port in ports]}])
+        assert out[0][dent]['rc'] == 0, 'Failed enslaving ports'
+
+    # 3. Set link up on all participant ports, bridges
+    out = await IpLink.set(input_data=[{dent: [{
+        'device': port,
+        'operstate': 'up'} for port in loopback_ports.values()]}])
+    assert out[0][dent]['rc'] == 0, 'Failed setting loopback links to state up'
+    out = await IpLink.set(input_data=[{dent: [{'device': bridge, 'operstate': 'up'} for bridge in bridges]}])
+    assert out[0][dent]['rc'] == 0, 'Failed setting bridge to state up'
+
+    await asyncio.sleep(wait_time)
+    stp_ports = await get_port_states(dent, bridge_names)
+
+    # 4. Bring down the indirect link. Verify the blocking port comes to forwarding after 51 seconds
+    out = await IpLink.set(input_data=[{dent: [{
+        'device': stp_ports['indirect_link'],
+        'operstate': 'down'}]}])
+    assert out[0][dent]['rc'] == 0, 'Failed setting interface to state down'
+    await asyncio.sleep(indirect_timeout)
+
+    out = await IpLink.show(input_data=[{dent: [
+        {'device': stp_ports['blocked_link'],
+         'cmd_options': '-j -d'}]}], parse_output=True)
+    assert out[0][dent]['rc'] == 0, 'Failed to get port detail'
+    err_msg = f'Port : {stp_ports["blocked_link"]} has to be in forwarding state'
+    assert out[0][dent]['parsed_output'][0]['linkinfo']['info_slave_data']['state'] == 'forwarding', err_msg
+
+    # 5. Bring back the indirect link and wait for the topology to re-converge.
+    out = await IpLink.set(input_data=[{dent: [{
+        'device': stp_ports['indirect_link'],
+        'operstate': 'up'}]}])
+    assert out[0][dent]['rc'] == 0, 'Failed setting interface to state up'
+
+    # 6. Bring down the direct link. Verify the blocking port comes to forwarding after 31 seconds
+    out = await IpLink.set(input_data=[{dent: [{
+        'device': stp_ports['direct_link'],
+        'operstate': 'down'}]}])
+    assert out[0][dent]['rc'] == 0, 'Failed setting interface to state down'
```

Comment:
Can be combined into one API call

Reply:
It can, but it 's two different tests steps, and think better to have them seperate

Reply:
```python
out = await IpLink.set(input_data=[{
    dent: [
        {'device': stp_ports['direct_link'], 'operstate': 'down'},
        {'device': stp_ports['indirect_link'], 'operstate': 'up'},
    ]}
])
assert out[0][dent]['rc'] == 0, 'Failed changing interface operstate'
```

Reply:
This is a bit easier to read. And we don't care as much about the assert message. It can be just 'ip link set failed'.

Reply:
hm..